src/app/child-dev-project/reporting/query.service.ts

```diff
@@ -0,0 +1,213 @@
+   * @param prefix the prefix which should be added to the string
+   * @returns a list where every string has the prefix
+   */
+  private addPrefix(ids: string[], prefix: string): string[] {
+    return ids.map((id) => (id.startsWith(prefix) ? id : prefix + ":" + id));
+  }
+
+  /**
+   * Creates an array containing the value of each key of the object.
+   * e.g. `{a: 1, b: 2} => [1,2]`
+   * This should be used when iterating over all documents of a given entity type because they are stored as
+   * `"{entity._id}": {entity}`
+   * @param obj the object which should be transformed to an array
+   * @returns the values of the input object as a list
+   */
+  private toArray(obj): any[] {
+    return Object.values(obj);
+  }
+
+  /**
+   * Returns a copy of the input array without duplicates
+   * @param data the array where duplicates should be removed
+   * @returns a list without duplicates
+   */
+  private unique(data: any[]): any[] {
+    return new Array(...new Set(data));
+  }
+
+  private count(data: any[]): number {
+    return data ? data.length : 0;
+  }
+
+  /**
+   * Turns a list of ids (with the entity prefix) into a list of entities
+   * @param ids the array of ids with entity prefix
+   * @returns a list of entity objects
+   */
+  private toEntities(ids: string[]): Entity[] {
+    return ids.map((id) => {
+      const prefix = id.split(":")[0];
+      return this.entities[prefix][id];
+    });
+  }
+
+  /**
+   * Returns all entities which reference a entity from the passed list of entities (by their id)
+   * @param srcEntities the entities for which relations should be found
+   * @param entityType the type of entities where relations should be looked for
+   * @param relationKey the name of the attribute that holds the reference.
+   *                    The attribute can be a string or a list of strings
+   * @returns a list of the related unique entities
+   */
+  private getRelated(
+    srcEntities: Entity[],
+    entityType: string,
+    relationKey: string
+  ): Entity[] {
+    const targetEntities = this.toArray(this.entities[entityType]);
+    const srcIds = srcEntities.map((entity) => entity.getId());
+    if (
+      targetEntities.length > 0 &&
+      Array.isArray(targetEntities[0][relationKey])
+    ) {
+      return targetEntities.filter((entity) =>
+        (entity[relationKey] as Array<string>).some((id) =>
+          srcIds.includes(id.split(":").pop())
+        )
+      );
+    } else {
+      return targetEntities.filter((entity) =>
+        entity[relationKey]
+          ? srcIds.includes(entity[relationKey].split(":").pop())
+          : false
+      );
+    }
+  }
+
+  private filterByObjectAttribute(
+    objs: any[],
+    attr: string,
+    key: string,
+    value: string
+  ): any[] {
+    const values = value.replace(new RegExp(" ", "g"), "").split("|");
+    return objs.filter((obj) => {
+      if (obj?.hasOwnProperty(attr)) {
+        return values.includes(obj[attr][key]?.toString());
```

Comment:
shouldn't this be an exact match rather than an "includes" here, when used for configurable-enum IDs?

Reply:
ah, never mind - the `value` parameter can have multiple options ...  maybe line 164 (the regex split) or the `values` variable can make that a bit more clear.

maybe add a jsdoc for the `value` parameter explaining `a value or a list of values separated by "|" (e.g. "ID_1|ID_2")`